ktor-features/ktor-auth-jwt/src/io/ktor/auth/jwt/JWTAuth.kt

```diff
@@ -0,0 +1,99 @@
+package io.ktor.auth.jwt
+
+import com.auth0.jwk.*
+import com.auth0.jwt.*
+import com.auth0.jwt.algorithms.*
+import com.auth0.jwt.exceptions.*
+import com.auth0.jwt.impl.*
+import com.auth0.jwt.interfaces.*
+import io.ktor.application.*
+import io.ktor.auth.*
+import io.ktor.http.*
+import io.ktor.response.*
+import java.security.interfaces.*
+import java.util.*
+
+private val JWTAuthKey: Any = "JWTAuth"
+
+data class JWTCredential(val payload: Payload) : Credential
+data class JWTPrincipal(val payload: Payload) : Principal
+
+fun AuthenticationPipeline.jwtAuthentication(verifier: JWTVerifier, validate: (JWTCredential) -> Principal?) {
+    intercept(AuthenticationPipeline.RequestAuthentication) { context ->
+        val token = call.getAuthToken()
+        val principal = verifyAndValidate(verifier, token, validate)
+        evaluate(token, principal, context)
+    }
+}
+
+fun AuthenticationPipeline.jwtAuthentication(jwkProvider: JwkProvider, issuer: String, validate: (JWTCredential) -> Principal?) {
+    intercept(AuthenticationPipeline.RequestAuthentication) { context ->
+        val token = call.getAuthToken()
+        val verifier = getVerifier(jwkProvider, issuer, token)
+        val principal = verifyAndValidate(verifier, token, validate)
+        evaluate(token, principal, context)
+    }
+}
+
+private suspend fun evaluate(token: String?, principal: Principal?, context: AuthenticationContext) {
+    val cause = when {
+        token == null -> NotAuthenticatedCause.NoCredentials
+        principal == null -> NotAuthenticatedCause.InvalidCredentials
+        else -> null
+    }
+    if (cause != null) {
+        context.challenge(JWTAuthKey, cause) {
+            call.respond(HttpStatusCode.Unauthorized)
```

Comment:
Shouldn't we provide 'WWW-Authenticate` header with `Bearer` challenge?

```
context.challenge(JWTAuthKey, NotAuthenticatedCause.InvalidCredentials) {
                it.success()
                call.respond(UnauthorizedResponse(HttpAuthHeader.bearerAuthChallenge(realm)))
            }

private fun HttpAuthHeader.Companion.bearerAuthChallenge(realm: String): HttpAuthHeader = 
    HttpAuthHeader.Parameterized("Bearer", mapOf(HttpAuthHeader.Parameters.Realm to realm))
```

You also can read https://www.scottbrady91.com/Kotlin/JSON-Web-Token-Verification-in-Ktor-using-Kotlin-and-Java-JWT

Also I am aware of parameters `error` and `error_description` that could be used to describe `cause` better so we can send only real if no credentials and error if the credentials are invalid.

Reply:
Regarding the `error` parameters: Do you mean as part of the Realm header?